### PR TITLE
refactor(azure): Extract SSH key cert handling to new certs module

### PR DESCRIPTION
**Pull request by peytonr18**
## Proposed Commit Message
```
refactor(azure): Extract SSH key cert handling to new certs module

Refactor x509 certificate and SSH key handling logic from 
OpenSSLManager and DataSourceAzure into a new dedicated 
module cloudinit/sources/azure/certs.py with three utility functions:

  - is_openssh_formatted(): Validates OpenSSH format keys
  - is_x509_certificate(): Validates x509 certificates via opensslprobert
  - convert_x509_to_openssh(): Converts x509 certs to OpenSSH keys

This refactoring maintains existing functionality without changes while 
preparing the codebase for upcoming work to support x509 certificates 
from Azure IMDS.
```

## Additional Context
**Purpose:**  
Refactors x509 certificate and SSH key handling logic from `OpenSSLManager` and `DataSourceAzure` into a dedicated module (`cloudinit/sources/azure/certs.py`). 

**Changes:**
- New module: `cloudinit/sources/azure/certs.py` with three functions:
 - `is_openssh_formatted()` - Validates OpenSSH format keys
 - `is_x509_certificate()` - Validates x509 certificates using openssl
 - `convert_x509_to_openssh()` - Converts x509 certs to OpenSSH keys
- Updated: `cloudinit/sources/helpers/azure.py` --> `_get_ssh_key_from_cert()` now uses new module
- Updated: `cloudinit/sources/DataSourceAzure.py` --> `_key_is_openssh_formatted()` now uses new module

**Test Coverage:**
- Unit tests for OpenSSH key validation and basic checks (various types, edge cases, markers, empty strings)
- Integration tests with `openssl` for x509 validation & with `openssl` + `ssh-keygen` for cert conversion
- Mocked tests for error handling (openssl/ssh-keygen failures)

## Merge type
- [x] Squash merge using "Proposed Commit Message"
---

Replaces #6561

### CODE REVIEW

File: cloudinit/sources/azure/certs.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    This checks if a given string is a valid OpenSSH public key format
+    (e.g., ssh-rsa, ssh-ed25519, ecdsa-sha2-nistp256).
+
+    """
+    if not key:
+        LOG.debug("Empty key content provided.")
+        return False
+    # See https://bugs.launchpad.net/cloud-init/+bug/1910835
+    if "\r\n" in key.strip():
```
Comment (cjp256) [inline cloudinit/sources/azure/certs.py]: what happens with AuthKeyLineParser if \r\n is found in the middle of the key? does it still validate?

File: cloudinit/sources/azure/certs.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+def is_openssh_formatted(key: str) -> bool:
+    """Validate whether or not the key is OpenSSH-formatted.
+
+    This checks if a given string is a valid OpenSSH public key format
+    (e.g., ssh-rsa, ssh-ed25519, ecdsa-sha2-nistp256).
+
+    """
+    if not key:
+        LOG.debug("Empty key content provided.")
```
Comment (cjp256) [inline cloudinit/sources/azure/certs.py]: References to "key" in the logs are a bit ambiguous.   Maybe "SSH key"?

File: cloudinit/sources/helpers/azure.py [hunk 3]
```diff
@@ -539,19 +535,27 @@ def parse_certificates(self, certificates_xml):
...
+        while True:
+            certificate = certs.extract_x509_certificate(remaining_data)
+            if certificate is None:
+                break
+
+            ssh_key = self._get_ssh_key_from_cert(certificate)
+            fingerprint = self._get_fingerprint_from_cert(certificate)
+            keys[fingerprint] = ssh_key
+
+            cert_index = remaining_data.find(certificate)
```
Comment (cjp256) [inline cloudinit/sources/helpers/azure.py]: I'm not clear on what is happening here?

I'm wondering if extract_x509_certificate() should be extract_x509_certificates() if multiple certs can be found, then we iterate through that set?

Comment (peytonr18) [reply cloudinit/sources/helpers/azure.py]: Thank you for calling this out - the loop was super hard to follow!

I refactored this to make the flow more explicit: 
- `extract_x509_certificate` is now `extract_x509_certificates` and returns all `certs (List[str])` from the bundle.
- `parse_certificates()` now just iterates that list and processes each certificate directly.

This removes the manual slicing loop and makes the “extract all certs, then process each” behavior much clearer.

**Review by bpryan99 — APPROVED** [review]
LGTM! Good test coverage.

File: cloudinit/sources/azure/certs.py (hunk 1 above)
Comment (peytonr18) [reply cloudinit/sources/azure/certs.py]: Edit: I've updated in the latest commit: instead of rejecting keys with embedded `\r\n`, we now sanitize them by stripping the CRLF sequences before validation. This handles the Azure-generated key case (LP: #1910835) where `\r\n` gets injected into the base64 data.

The flow in `_get_public_ssh_keys_from_imds()` is now:
1. Collect raw keys from IMDS
2. Sanitize any keys containing `\r\n` via `sanitize_openssh_key()`
3. Validate with `is_openssh_formatted()` 
4. Return clean, validated keys

For example, a key like `ssh-rsa AAAA\r\nBBBB comment` gets sanitized to `ssh-rsa AAAABBBB comment`, validated, and returned cleanly for `authorized_keys`.

File: cloudinit/sources/azure/certs.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+LOG = logging.getLogger(__name__)
+
+_CERTIFICATE_BLOCK_RE = re.compile(
+    r"-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----",
+    re.DOTALL,
+)
+
+
+def sanitize_openssh_key(key: str) -> str:
+    r"""Sanitize an OpenSSH key by removing embedded CRLF sequences.
```
Comment (cjp256) [inline cloudinit/sources/azure/certs.py:20]: r""" ?

File: cloudinit/sources/azure/certs.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    Azure-generated SSH keys may contain \\r\\n sequences embedded in the
```
Comment (cjp256) [inline cloudinit/sources/azure/certs.py]:
```suggestion
    Azure-generated SSH keys may contain CRLF (\r\n) sequences embedded in the
```

File: cloudinit/sources/azure/certs.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    base64 key data. This strips those sequences so the key can be properly
+    parsed and written to authorized_keys.
+
+    See https://bugs.launchpad.net/cloud-init/+bug/1910835
```
Comment (cjp256) [inline cloudinit/sources/azure/certs.py:29]: add a note that Azure doesn't do this today, but it's possible these keys are still in use

File: cloudinit/sources/DataSourceAzure.py [hunk 7]
```diff
@@ -1044,6 +1044,11 @@ def _get_public_keys_from_imds(self, imds_md: dict) -> List[str]:
             report_diagnostic_event(log_msg, logger_func=LOG.debug)
             raise
 
+        ssh_keys = [
+            certs.sanitize_openssh_key(key) if "\r\n" in key else key
```
Comment (cjp256) [inline cloudinit/sources/DataSourceAzure.py]: sanitize_openssh_key() is fine to call on for all keys since it does the check

File: cloudinit/sources/DataSourceAzure.py [hunk 8]
```diff
@@ -1745,17 +1750,7 @@ def _key_is_openssh_formatted(key):
...
-        return False
-
-    parser = ssh_util.AuthKeyLineParser()
-    try:
-        akl = parser.parse(key)
-    except TypeError:
-        return False
-
-    return akl.keytype is not None
+    return certs.is_openssh_formatted(key)
```
Comment (cjp256) [inline cloudinit/sources/DataSourceAzure.py]: this function can be removed in favor of this call

File: tests/unittests/sources/azure/test_certs.py [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+    """Skip the test if any required external command is unavailable."""
+    for command in commands:
+        if subp.which(command) is None:
+            pytest.skip(f"{command} not available")
+
+
+class TestIsOpensshFormatted:
+    """Test is_openssh_formatted() function."""
+
+    def test_valid_rsa_key_without_comment(self):
```
Comment (cjp256) [inline tests/unittests/sources/azure/test_certs.py]: use pytest.mark.parametrize() to consolidate most of these tests

Comment (cjp256) [reply tests/unittests/sources/azure/test_certs.py]: at least into a test_valid() and test_invalid() set

File: tests/unittests/sources/test_azure_helper.py [hunk 10]
```diff
@@ -556,6 +556,78 @@ def test_parse_certificates(self, mock_decrypt_certs):
...
+            return_value=[cert1, cert2],
+        ):
+            sslmgr = azure_helper.OpenSSLManager()
+            result = sslmgr.parse_certificates("fake-xml")
+
+        assert result == {"fp1": "ssh-cert-1", "fp2": "ssh-cert-2"}
+        m_decrypt.assert_called_once_with("fake-xml")
+        assert m_get_key.call_args_list == [
+            mock.call(
+                "-----BEGIN CERTIFICATE-----\r\n"
```
Comment (cjp256) [inline tests/unittests/sources/test_azure_helper.py]: reuse the vars

```suggestion
                cert1
```

and so on

File: cloudinit/sources/azure/certs.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    attempting to parse it with openssl.
+    """
+    if not cert:
+        LOG.debug("Empty certificate provided.")
+        return False
+
+    cert = cert.strip()
+
+    if "-----BEGIN CERTIFICATE-----" not in cert:
+        LOG.debug("Missing BEGIN CERTIFICATE marker.")
```
Comment (cjp256) [inline cloudinit/sources/azure/certs.py]: `Missing` somewhat implies that it should be there. If `is_x509_certificate()` is called to identify it, it's more like

```suggestion
        LOG.debug("No BEGIN CERTIFICATE marker.")
```

Fwiw I noticed I'm guilty of the same thing over in OvfEnvXml

File: cloudinit/sources/azure/certs.py (hunk 4 above)
Comment (peytonr18) [reply cloudinit/sources/azure/certs.py:20]: The reason for this was that the format CI test throws an error and fails without an r ---

```
D301 Use `r"""` if any backslashes in a docstring
  --> cloudinit/sources/azure/certs.py:20:5
   |
19 |   def sanitize_openssh_key(key: str) -> str:
20 | /     """Sanitize an OpenSSH key by removing embedded CRLF sequences.
21 | |
22 | |     Azure-generated SSH keys may contain CRLF (\\r\\n) sequences embedded in the
23 | |     base64 key data. This strips those sequences so the key can be properly
24 | |     parsed and written to authorized_keys.
25 | |
26 | |     Note: Azure no longer generates keys with embedded CRLF sequences,
27 | |     but previously provisioned keys may still be in use.
28 | |
29 | |     See https://bugs.launchpad.net/cloud-init/+bug/1910835
30 | |     """
   | |_______^
31 |       key = key.strip()
32 |       if "\r\n" in key:
   |
help: Add `r` prefix
```

Comment (cjp256) [reply cloudinit/sources/azure/certs.py:20]: interesting.  in that case I think \\r\\n only needs to be \r\n?

File: cloudinit/sources/azure/certs.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+import logging
+import re
+from typing import List, Optional
+
+from cloudinit import ssh_util, subp
+
+LOG = logging.getLogger(__name__)
+
+_CERTIFICATE_BLOCK_RE = re.compile(
+    r"-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----",
```
Comment (cjp256) [inline cloudinit/sources/azure/certs.py:14]: this regex seems stricter than the existing logic, but i can't think of a reason this is a problem. looks like there is a TRUSTED CERTIFICATE that seems unlikely to be relevant

File: tests/unittests/sources/azure/test_certs.py [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+        assert certs.is_x509_certificate(value) is False
+
+    def test_invalid_certificate_content(self, monkeypatch):
+        """Certificate with invalid content should return False."""
+
+        def _raise(*args, **kwargs):
+            raise subp.ProcessExecutionError("unable to load certificate")
+
+        monkeypatch.setattr("cloudinit.sources.azure.certs.subp.subp", _raise)
+        assert certs.is_x509_certificate(_INVALID_X509_CERT) is False
```
Comment (cjp256) [inline tests/unittests/sources/azure/test_certs.py:163]: how do we know it's false for the reasons you expect?

File: tests/unittests/sources/azure/test_certs.py [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+
+    @pytest.mark.allow_subp_for("openssl")
+    def test_valid_certificate_integration(self, cert_data):
+        """Integration test: Actually validate certificate with openssl."""
+        _require_commands("openssl")
```
Comment (cjp256) [inline tests/unittests/sources/azure/test_certs.py]: @pytest.mark.skipif would be best here

File: tests/unittests/sources/azure/test_certs.py [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+        if subp.which(command) is None:
+            pytest.skip(f"{command} not available")
+
+
+@pytest.fixture
+def data_file_path():
+    """Return a helper that resolves Azure test data file paths."""
+
+    def _path(name):
+        return os.path.join("tests", "data", "azure", name)
```
Comment (cjp256) [inline tests/unittests/sources/azure/test_certs.py]: would be a little nicer if using pathlib.Path

File: tests/unittests/sources/azure/test_certs.py [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+        results = iter([False, True])
+        monkeypatch.setattr(
+            "cloudinit.sources.azure.certs.is_x509_certificate",
+            lambda cert: next(results),
+        )
+
+        result = certs.extract_x509_certificates(bundle)
+
+        assert len(result) == 1
+        assert "VALID" in result[0]
```
Comment (cjp256) [inline tests/unittests/sources/azure/test_certs.py]: VALID is also in INVALID

Comment (cjp256) [reply tests/unittests/sources/azure/test_certs.py]: assert result == [valid_cert] ?

File: tests/unittests/sources/azure/test_certs.py [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+        bundle = cert1 + "\n" + cert2
+
+        monkeypatch.setattr(
+            "cloudinit.sources.azure.certs.is_x509_certificate",
+            lambda cert: True,
+        )
+
+        result = certs.extract_x509_certificates(bundle)
+
+        assert len(result) == 2
```
Comment (cjp256) [inline tests/unittests/sources/azure/test_certs.py]: expand all these into assert result == ...

File: tests/unittests/sources/azure/test_certs.py (hunk 13 above)
Comment (peytonr18) [reply tests/unittests/sources/azure/test_certs.py:163]: In the latest commit, I added `m_subp.assert_called_once()` after the assertion. The test cert includes the BEGIN/END markers so it passes the guard clauses --- `assert_called_once()` now confirms that it actually reached the openssl validation path and the `False` is from the `ProcessExecutionError`!

**Review by anhvoms — APPROVED** [review]

**Review by cjp256 — APPROVED** [review]
LGTM, thanks Peyton!

**Comment by cjp256** [comment]
@blackboxsw this is ready for review when you get a chance, thanks!

File: cloudinit/sources/azure/certs.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
+# Copyright (C) 2024 Microsoft Corporation.
```
Comment (blackboxsw) [inline cloudinit/sources/azure/certs.py]: Shall we represent the current year, since we are mentioning the Copyright anyway?

**Comment by peytonr18** [comment]
Hi @blackboxsw, this one is on track to be marked stale whenever you get a chance - thank you!

**Comment by peytonr18** [comment]
@blackboxsw, asking to remove stale tag, PR awaiting review.

**Comment by cjp256** [comment]
@blackboxsw this probably got buried and missed :)

**Comment by cjp256** [comment]
@holmanb can you give this a review when you get a chance?  would appreciate it!  thanks!

**Review by holmanb — APPROVED** [review]
This looks good to me. Thanks!
